daemon: add option to dynamically size BPF maps based on system memory

tklauser opened the pull request:
Add the new `bpf-map-dynamic-size` option which will size the large BPF
maps (currently CT TCP and non-TCP, NAT and Policy) based on the
total system memory.

If enabled, approximately 1% of the total system memory will be used to
allocate the BPF maps. This factor was chosen heuristically such that on
a node with 4GB of memory, the maps will have roughly the default size.
The amount of memory will be distributed

If the user specifies any of the map sizes explicitly using an option,
that value will take precedence over the dynamically determined size for
that particular map, i.e. when using

```
cilium-agent --bpf-map-dynamic-size --bpf-ct-global-tcp-max 100000
```

the TCP conntrack BPF map will be allocated for a maximum of 100000
entries while the non-TCP conntrack, NAT and Policy maps will have their
size determined dynamically based on total system memory.

As with the manually specified sizes, the dynamically determined map
sizes are always capped by respective minimum and maximum number of
entires for each map.

Updates #10056

```release-note
Add command line option to dynamically size BPF maps based on total system memory.
```

qmonnet reviewed (commented):
Looks good, but it would maybe benefit from more documentation? Having just the cmdref mentioning the option does not tell anything about the amount of memory used by the heuristic, for example. Also maybe update the `BPF Map Limitations` section in `Documentation/architecture.rst`?

pchaigno commented on daemon/cmd/daemon_main.go:
```diff
@@ -674,6 +674,9 @@ func init() {
 	flags.Int(option.PolicyMapEntriesName, defaults.PolicyMapEntries, "Maximum number of entries in endpoint policy map (per endpoint)")
 	option.BindEnv(option.PolicyMapEntriesName)
 
+	flags.Bool(option.MapEntriesGlobalDynamicSizeName, false, "Enable dynamic sizing of BPF maps based on total system memory")
```
The option description (or the doc.) should probably state which maps are affected by this option.

In general, if a user wants to reduce/increase the memory usage after switching dynamic sizing on, do they have a way to know the current sizes?

tklauser replied:
> The option description (or the doc.) should probably state which maps are affected by this option.

Good point, will adjust the usage text.

> In general, if a user wants to reduce/increase the memory usage after switching dynamic sizing on, do they have a way to know the current sizes?

The debug logs will print the calculated map sizes, though there won't be a number in terms of system memory. but users could still use `bpftool` to get the key and value size and estimate the map size using these numbers.

tgraf commented on install/kubernetes/cilium/values.yaml:
```diff
@@ -219,7 +219,10 @@ global:
 
-    # montiorAggregation is the level of aggregation for datapath trace events
+    # mapDynamicSize enables dynamic sizing of BPF maps based on available system memory
+    mapDynamicSize: false
```
I think we should enable this by default. The purpose of this is to consume less memory by default. We can keep it disabled for existing deployments and enable it for new deployment by enabling it via the helm values here.

aanm replied:
Not addressed?

tklauser replied:
I addressed it in my local branch but didn't push yet due to the issues I found regarding the correct ratios for the individual maps.

tklauser replied:
Now addressed and set to a ratio of `0.03` (3%) by default which roughly corresponds to the default map sizes on a node with 4 GiB of memory.

This only affects the helm value, not the default value for the command line option (which will set a ratio of `0.0`, i.e. disable the dynamic sizing.

pchaigno commented on pkg/option/config.go:
```diff
@@ -2143,6 +2144,90 @@ func (c *DaemonConfig) Populate() {
+		// provided by the user will override the calculated value and
+		// also the max. There will be check for maximum size later on
+		// in DaemonConfig.Validate()
+		memoryAvailableForMaps := int(vms.Total / 100)
```
Should the percentage of memory used for maps be configurable? I guess we can also expose it later if there's demand.

tklauser replied:
Good point. Initially I was afraid that this would give users a way to easily shot themselves in the foot with an OOM, but given that we cap the min/max of the map entries, I think it would be safe to allow a percentage here (along with sufficient documentation on how to use it, e.g. with some example calculations as suggested in https://github.com/cilium/cilium/pull/10780#discussion_r404199537). And given that we enable this by default as @tgraf suggested, we could convert the option to take a percentage instead of a boolean.

soumynathan replied:
That would be good added value for the users if we provide the information.

tklauser replied:
Changed the command line option to take a ratio and and this is now used instead of the fixed value of 1%

tgraf commented on pkg/option/config.go:
```diff
@@ -2143,6 +2144,90 @@ func (c *DaemonConfig) Populate() {
+	c.NATMapEntriesGlobal = viper.GetInt(NATMapEntriesGlobalName)
+	c.PolicyMapMaxEntries = viper.GetInt(PolicyMapEntriesName)
+
+	if viper.GetBool(MapEntriesGlobalDynamicSizeName) {
```
Can you refactor this into a calculation function so we can unit test the calculation?

tgraf commented on pkg/option/config.go:
```diff
@@ -2143,6 +2144,90 @@ func (c *DaemonConfig) Populate() {
+		// provided by the user will override the calculated value and
+		// also the max. There will be check for maximum size later on
+		// in DaemonConfig.Validate()
+		memoryAvailableForMaps := int(vms.Total / 100)
```
Can you add some calculation examples for memory 500MB, 1GB, 4GB, 16GB?

pchaigno reviewed (commented):
Nice PR!

soumynathan reviewed (commented):
Nit: In the commit message, there is a typo.
'This factor was chosen heuristically such that on
a node with 4GB or memory, '.
Probably it should say, 'a node with 4GB  of memory' or '4GP or more memory'.

tklauser committed:
daemon: add option to dynamically size BPF maps based on system memory

Add the new `bpf-map-dynamic-size-ratio` option which will size the
large BPF maps (currently CT TCP and non-TCP, NAT and policy) based on
the total system memory.

If set to a value > 0.0, the given ratio of the total system memory will
be used to allocate the BPF maps. The default scaling was chosen
heuristically such that on a node with 4GiB of total system memory, the
maps will have roughly the default size when using a ratio of 0.03 (3%).
The amount of memory will be distributed among the TCP CT, non-TCP CT,
NAT and policy tables proportional to their default sizes.

If the user specifies any of the map sizes explicitly using an option,
that value will take precedence over the dynamically determined size for
that particular map, i.e. when using

```
cilium-agent --bpf-map-dynamic-size-ratio 0.03 --bpf-ct-global-tcp-max 100000
```

the TCP conntrack BPF map will be allocated for a maximum of 100000
entries while the non-TCP conntrack, NAT and policy maps will have their
size determined dynamically based on total system memory.

As with the manually specified sizes, the dynamically determined map
sizes are always clamped to respective minimum and maximum number of
entires for each map.

Updates #10056

Signed-off-by: Tobias Klauser <tklauser@distanz.ch>

aanm reviewed (commented):
This sounds really great. couple questions / comments:
1) I assume we will only enable this by default for new installations and not for upgrades.
2) Your PR descriptions states 1% of memory, is that a typo? 1% looks extremely low.
3) From a usability perspective, and this can be addressed as a follow up, I would love to see an option where the user wouldn't care about map element sizes and the she would only define the maximum amount of memory BPF maps would take, for example: `--bpf-max-memory=1GB`. This would simplify usability since the user does not need to worry how much memory does 800k TCP CT entries will take in her system for example. Cilium would translate how much 1GB would be for each bpf map in terms of number of entries.

qmonnet commented on Documentation/architecture.rst:
```diff
@@ -258,6 +258,13 @@ Tunnel                   node             64k             Max 32k nodes (IPv4+IP
+line options for ``cilium-agent``. A given capacity can be set using
+``--bpf-ct-global-tcp-max``, ``--bpf-ct-global-any-max``,
+``--bpf-nat-global-max`` and ``--bpf-policy-map-max``. Using the
+``--bpf-map-dynamic-size-ratio`` these upper capacity limits determined based on
```
Now `--bpf-fragments-map-max` as well.

tklauser replied:
Will add it, thanks. Do you think it would make sense to dynamically scale the fragments map as well?

qmonnet replied:
I've been wondering, too. I think we don't expect too many fragmented packets anyway, so it's probably not worth the trouble at this time. Worst case, users can still use the option to set the size manually. So let's leave it out for now?

tklauser replied:
Agree, let's leave it out for now. Also given that it's default size is fairly small. And if the need arises we can still dynamically size it in a follow-up PR.

qmonnet commented on pkg/option/config.go:
```diff
@@ -2735,6 +2711,160 @@ func (c *DaemonConfig) populateHostServicesProtos() error {
+		c.calculateDynamicBPFMapSizes(vms.Total, dynamicSizeRatio)
+	} else if dynamicSizeRatio < 0.0 {
+		return fmt.Errorf("specified dynamic map size ratio %f must be ≥ 0.0", dynamicSizeRatio)
+	}
```
What happens if `dynamicSizeRatio > 1.0`? No error message?

tklauser replied:
Good point. The comparison above is ` 0.0 < dynamicSizeRatio && dynamicSizeRatio <= 1.0` so a `dynamicSizeRatio > 1.0` at least not lead to a wrong calculation. Will adjust the condition and the error message accordingly.

qmonnet commented on pkg/option/config.go:
```diff
@@ -2735,6 +2711,160 @@ func (c *DaemonConfig) populateHostServicesProtos() error {
+	//  Policy:             8     + 24    = 32
+	sizeofCTEntry := 94
+	sizeofNATEntry := 88
+	sizeofPolicyEntry := 32
```
Any chance we could compute those from the sizes of the actual structs?

tklauser replied:
Unfortunately, these structs are defined in `pkg/maps` which in turn depends on `pkg/option` which would lead to a circular dependency, which I'd like to avoid f possible. I agree that specifying these sizes manually is far from optimal but then again the point is not to get them exactly right, but to have them just approximately right so the available memory will be distributed fairly.

qmonnet replied:
I thought we might hit a circular dependency but wasn't sure. Ok, no problem, thanks for looking into it!

aanm replied:
@tklauser we can add these values as fields part of the `DaemonConfig` and set them before calling `Populate`, here
https://github.com/cilium/cilium/blob/93b209b9342fb2901b95358f6ad11f176cfaed70/daemon/cmd/daemon_main.go#L799-L801

tklauser replied:
Thanks, this turned up that I was actually underestimating the NAT map entry size!

qmonnet reviewed (approved):
Looks good, thanks for the changes!

pchaigno commented on Documentation/architecture.rst:
```diff
@@ -259,6 +259,14 @@ Tunnel                   node             64k             Max 32k nodes (IPv4+IP
+``--bpf-nat-global-max``, ``--bpf-policy-map-max``, and
+``--bpf-fragments-map-max``. Using ``--bpf-map-dynamic-size-ratio`` these
+upper capacity limits are determined based on the given ratio of the total
+system memory.
```
This sentence might be a bit misleading. `--bpf-map-dynamic-size-ratio` doesn't compute a limit for the fragment map mentioned in the previous sentence.

tklauser replied:
Good point. I'll wait whether there are any other review comments and address this in batch.

tklauser replied:
Now fixed. Also took the opportunity to slightly extend the paragraph about `--bpf-map-dynamic-size-ratio`.

pchaigno reviewed (approved):
One nit below. LGTM otherwise!

aanm commented on Documentation/cmdref/cilium-agent.md:
```diff
@@ -32,6 +32,7 @@ cilium-agent [flags]
       --bpf-ct-timeout-service-any duration           Timeout for service entries in non-TCP CT table (default 1m0s)
       --bpf-ct-timeout-service-tcp duration           Timeout for established service entries in TCP CT table (default 6h0m0s)
       --bpf-fragments-map-max int                     Maximum number of entries in fragments tracking map (default 8192)
+      --bpf-map-dynamic-size-ratio float              Ratio (0.0-1.0) of total system memory to use for dynamic sizing of CT, NAT and policy BPF maps. Set to 0.0 to disable dynamic BPF map sizing (default: 0.0)
```
Does this conflicts with the `preallocate-bpf-maps` option?

tklauser replied:
No, `--bpf-map-dynamic-size-ratio` only influences the number of entries in a map (like the `--bpf-*-map-max` flags) while `preallocate-bpf-maps` leads to the BPF hash maps being created without the `BPF_F_NO_PREALLOC` flag so they are preallocated by the kernel.

joestringer reviewed (commented):
Please document the upgrade implications (at minimum the "new ConfigMap options" section of the upgrade guide) either in this PR or a follow-up.

joestringer commented on Documentation/install/upgrade.rst:
```diff
@@ -434,6 +434,24 @@ Upgrading from >=1.7.0 to 1.8.y
+    manually using the ``ctTcpMax``, ``ctAnyMax``, ``natMax`` options will take
+    precedence over the dynamically determined value.
+
+    On upgrades of existing installations, this option is disable by default,
```
```suggestion
    On upgrades of existing installations, this option is disabled by default,
```

joestringer reviewed (approved):
Minor nit, feel free to follow up separately since it's a docs-only change and will not require full CI run to validate.
